src/write.rs

```diff
@@ -788,75 +785,37 @@ impl<W: Write + Seek> ZipWriter<W> {
+            let file = ZipFileData::initialize_local_block(
+                name,
+                &options,
+                raw_values,
                 header_start,
-                extra_data_start: None,
-                data_start: OnceLock::new(),
-                central_header_start: 0,
-                external_attributes: permissions << 16,
-                large_file: options.large_file,
-                aes_mode,
+                None,
                 aes_extra_data_start,
+                compression_method,
+                aes_mode,
+                extra_field,
+            );
 
-                extra_fields: Vec::new(),
-            };
             let index = self.insert_file_data(file)?;
             let file = &mut self.files[index];
             let writer = self.inner.get_plain();
-            // local file header signature
-            writer.write_u32_le(spec::LOCAL_FILE_HEADER_SIGNATURE)?;
-            // version needed to extract
-            writer.write_u16_le(file.version_needed())?;
-            // general purpose bit flag
-            let is_utf8 = std::str::from_utf8(&file.file_name_raw).is_ok();
-            let is_ascii = file.file_name_raw.is_ascii();
-            let flag = if is_utf8 && !is_ascii { 1u16 << 11 } else { 0 }
-                | if file.encrypted { 1u16 << 0 } else { 0 };
-            writer.write_u16_le(flag)?;
-            // Compression method
-            #[allow(deprecated)]
-            writer.write_u16_le(file.compression_method.to_u16())?;
-            // last mod file time and last mod file date
-            writer.write_u16_le(file.last_modified_time.timepart())?;
-            writer.write_u16_le(file.last_modified_time.datepart())?;
-            // crc-32
-            writer.write_u32_le(file.crc32)?;
-            // compressed size and uncompressed size
-            if file.large_file {
-                writer.write_u32_le(spec::ZIP64_BYTES_THR as u32)?;
-                writer.write_u32_le(spec::ZIP64_BYTES_THR as u32)?;
-            } else {
-                writer.write_u32_le(file.compressed_size as u32)?;
-                writer.write_u32_le(file.uncompressed_size as u32)?;
-            }
-            // file name length
-            writer.write_u16_le(file.file_name_raw.len() as u16)?;
-            // extra field length
-            let mut extra_field_length = file.extra_field_len();
-            if file.large_file {
-                extra_field_length += 20;
-            }
-            if extra_field_length + file.central_extra_field_len() > u16::MAX as usize {
-                let _ = self.abort_file();
-                return Err(InvalidArchive("Extra data field is too large"));
+
+            let block = match file.local_block() {
+                Ok(block) => block,
+                Err(e) => {
+                    let _ = self.abort_file();
+                    return Err(e);
+                }
+            };
```

Comment:
Extract lines 804-810, 811-815 and the other usages of this `match` pattern into a private or `pub(crate)` method of `ZipWriter` that takes a `FnOnce<&mut Self, ZipResult<T>>`.

Reply:
This won't work here as `file` cannot be captured into a closure (even an `FnOnce` closure) when mutably borrowed, e.g. (with the new method `self.do_or_abort_file()`:
```rust
error[E0499]: cannot borrow `*self` as mutable more than once at a time
   --> src/write.rs:812:25
    |
809 |             let file = &mut self.files[index];
    |                             ---------- first mutable borrow occurs here
...
812 |             let block = self.do_or_abort_file(|| file.local_block())?;
    |                         ^^^^                     ---- first borrow later captured here by closure
    |                         |
    |                         second mutable borrow occurs here
```

I found this failure to occur with the other attempts to use this pattern, so I'm not sure how to make this work.